### PR TITLE
DBZ-2662 Handle ChangeEventQueue by the size in bytes

### PR DESCRIPTION
**Pull request by vanhoale**
[JIRA](https://issues.redhat.com/browse/DBZ-2662)

### CODE REVIEW

File: debezium-core/pom.xml [hunk 1]
```diff
@@ -101,6 +101,12 @@
             <artifactId>kafka-connect-avro-converter</artifactId>
             <scope>test</scope>
         </dependency>
+        <!-- open libraries -->
+        <dependency>
+            <groupId>com.twitter.common</groupId>
+            <artifactId>objectsize</artifactId>
+            <version>0.0.12</version>
```
Comment (rk3rn3r) [inline debezium-core/pom.xml:108]: I think a newer version of the ObjectSizeCalculator is available here: https://github.com/twitter/util/blob/develop/util-core/src/test/java/com/twitter/util/ObjectSizeCalculator.java but it's not available on maven central and twitter only uses it for unit tests.

File: debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java [hunk 2]
```diff
@@ -154,6 +171,7 @@ public void enqueue(T record) throws InterruptedException {
                 metronome.pause();
                 LOGGER.debug("checking for more records...");
             }
+            currentQueueSizeInBytes.addAndGet(-ObjectSizeCalculator.getObjectSize(records));
```
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java:174]: @vanhoale Would it be possible to handle queue size set to zero as a special value that disables the rate limitng and calculation of the  object size?

Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java:174]: @vanhoale
`Field::isPositiveLong` would become `Field::isNonNegativeLong` and description will be updated, default value will be `0`
this line will be encapusalted in `If (maxQueueSizeInBytes > 0)`
So if the config value is not set or is set to zero the calculation will not be triggered. The waiting loop above will wait only for non zero maxQueueSizeInBytes

File: debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java [hunk 3]
```diff
@@ -126,9 +139,13 @@ public void enqueue(T record) throws InterruptedException {
         if (LOGGER.isDebugEnabled()) {
             LOGGER.debug("Enqueuing source record '{}'", record);
         }
-
+        // Waiting for queue to add more record.
+        while (currentQueueSizeInBytes.get() > maxQueueSizeInBytes) {
+            Thread.sleep(5000);
```
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java:144]: Instead of hardoced value I recommend to reuse `pollInterval`

**Review by vanhoale — COMMENTED** [review]
Started review

File: debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java (hunk 3 above)
Comment (vanhoale) [reply debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java:144]: Thanks! I will update it

File: debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java (hunk 2 above)
Comment (vanhoale) [reply debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java:174]: is it returned DEFAULT_MAX_QUEUE_SIZE_IN_BYTES if I don't pass the max.queue.size.in.bytes variable?

Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java:174]: Yes and we want to disable queue size limitng by default, hence choosign the 0.

Comment (vanhoale) [reply debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java:174]: ah, got your point, so you want a logic like: if we don't pass the maxQueueSizeInBytes then we shouldn't use ObjectSizeCalculator.getObjectSize() method?

Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java:174]: Exactly!

Comment (vanhoale) [reply debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java:174]: I like it, will update it

File: debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java [hunk 4]
```diff
@@ -242,6 +242,7 @@ public static BinaryHandlingMode parse(String value, String defaultValue) {
     public static final String DATABASE_CONFIG_PREFIX = "database.";
     private static final String CONVERTER_TYPE_SUFFIX = ".type";
     public static final long DEFAULT_RETRIABLE_RESTART_WAIT = 10000L;
+    public static final long DEFAULT_MAX_QUEUE_SIZE_IN_BYTES = 500 * 1000 * 1024L; // 500MB;
```
Comment (vanhoale) [inline debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java:245]: @jpechane should I set the default value to 0/nagative value, so I can check it to calculate the object size or not?

File: debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java [hunk 5]
```diff
@@ -242,6 +242,7 @@ public static BinaryHandlingMode parse(String value, String defaultValue) {
     public static final String DATABASE_CONFIG_PREFIX = "database.";
     private static final String CONVERTER_TYPE_SUFFIX = ".type";
     public static final long DEFAULT_RETRIABLE_RESTART_WAIT = 10000L;
+    public static final long DEFAULT_MAX_QUEUE_SIZE_IN_BYTES = -1; // In case we don't want to pass max.queue.size.in.bytes;
```
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java:245]: Why `-1` and not `0`?

File: debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java [hunk 6]
```diff
@@ -154,6 +174,9 @@ public void enqueue(T record) throws InterruptedException {
                 metronome.pause();
                 LOGGER.debug("checking for more records...");
             }
+            if (maxQueueSizeInBytes > 0) {
+                currentQueueSizeInBytes.addAndGet(-ObjectSizeCalculator.getObjectSize(records));
```
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java:178]: I don't think this is correct. IMHO sum of the object sizes does not equal the size of array of the same objects

Comment (vanhoale) [reply debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java:178]: @jpechane Confirmed! it's not equal, should I wrap the record in a key-pair object so I can get the key that can be referred to the object size later?

File: debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java (hunk 5 above)
Comment (vanhoale) [reply debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java:245]: I can update it to 0

**Comment by jpechane** [comment]
@vanhoale I think the core functionality is fine now. I'd like to ask you for few more improvements

- expose the current queue size ans max queue size as JMX metric, see `io.debezium.connector.base.ChangeEventQueueMetrics`
- with the feature above in please please add to one of the connectors a test that will will verify the size of the queue before events are written, when some of them are in it and that the size is zero again when all records were consumed
- add new option and metrics to the documentation
- remove the Maven dependency and pull the sourcecode into the project as mentioned elsewhere

Thanks a lot for this new feature

**Comment by vanhoale** [comment]
> @vanhoale I think the core functionality is fine now. I'd like to ask you for few more improvements
> 
> * expose the current queue size ans max queue size as JMX metric, see `io.debezium.connector.base.ChangeEventQueueMetrics`
> * with the feature above in please please add to one of the connectors a test that will will verify the size of the queue before events are written, when some of them are in it and that the size is zero again when all records were consumed
> * add new option and metrics to the documentation
> * remove the Maven dependency and pull the sourcecode into the project as mentioned elsewhere
> 
> Thanks a lot for this new feature

Thanks @jpechane ! Yes, I will add improvements.

**Comment by vanhoale** [comment]
@jpechane should I use caffeine for the CacheBuilder? As mentioned the bug in the code? I also read an article about the guava inspired by caffeine: https://www.linkedin.com/pulse/asynchronous-reload-guava-cache-caffeine-bhanuprakash-devapatla/

**Comment by vanhoale** [comment]
No, the caffein version is much slower then the https://github.com/twitter-archive/commons/blob/master/src/java/com/twitter/common/objectsize/ObjectSizeCalculator.java one

**Comment by ben-manes** [comment]
@vanhoale can you please explain what is slower? The use of CacheBuilder inside of `ObjectSizeCalculator` is an unbounded memoizing map, so technically it is unnecessary as of Java 8. That correlates to our [compute benchmark](https://github.com/ben-manes/caffeine/wiki/Benchmarks#compute), where Caffeine is faster that Guava or using `computeIfAbsent` directly. The latter only because we avoid the pessimistic internal lock by optimistically performing a lock-free `get(key)`. That same optimization could be applied to remove the cache dependencies.

Or was the comparison against something else?

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresMetricsIT.java [hunk 7]
```diff
@@ -190,6 +191,66 @@ private void assertStreamingMetrics() throws Exception {
...
+                .with(PostgresConnectorConfig.MAX_BATCH_SIZE, 1)
+                .with(PostgresConnectorConfig.POLL_INTERVAL_MS, 5000L)
+                .with(PostgresConnectorConfig.MAX_QUEUE_SIZE_IN_BYTES, 10L);
+        start(PostgresConnector.class, configBuilder.build());
+
+        waitForStreamingToStart();
+        for (int i = 0; i < recordCount - 1; i++) {
+            TestHelper.execute(INSERT_STATEMENTS);
+        }
+        Thread.sleep(1000);
```
Comment (jpechane) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresMetricsIT.java:243]: Test looks good but this is a brittle approach. PLease check `io.debezium.embedded.AbstractConnectorTest.waitForStreamingRunning(String, String, String)` code how to use `Awaitility` lib and use until to wait for CurrentQueueSizeInBytes > 0 and QueueRemainingCapacity == 9
Please follow the same approach in other tests

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresMetricsIT.java [hunk 8]
```diff
@@ -190,6 +191,66 @@ private void assertStreamingMetrics() throws Exception {
...
+        Assertions.assertThat(mBeanServer.getAttribute(getStreamingMetricsObjectName(), "CurrentQueueSizeInBytes")).isNotEqualTo(0L);
+        Thread.sleep(500);
+        Assertions.assertThat(mBeanServer.getAttribute(getStreamingMetricsObjectName(), "QueueRemainingCapacity")).isEqualTo(9);
+
+        SourceRecords records = consumeRecordsByTopic(recordCount);
+        Thread.sleep(1000);
```
Comment (jpechane) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresMetricsIT.java:249]: This should be unnecessary, the queue should already be of size 0 at time of all records being consumed

**Comment by jpechane** [comment]
@ben-manes @vanhoale Interesting discussion, the benchmark looks convincing. @vanhoale Could you please share your experiences?

**Comment by vanhoale** [comment]
> @vanhoale can you please explain what is slower? The use of CacheBuilder inside of `ObjectSizeCalculator` is an unbounded memoizing map, so technically it is unnecessary as of Java 8. That correlates to our [compute benchmark](https://github.com/ben-manes/caffeine/wiki/Benchmarks#compute), where Caffeine is faster that Guava or using `computeIfAbsent` directly. The latter only because we avoid the pessimistic internal lock by optimistically performing a lock-free `get(key)`. That same optimization could be applied to remove the cache dependencies.
> 
> Or was the comparison against something else?

Hi @ben-manes, thanks so much for looking at this, and for the benchmark chart that looks nice!

I have done 2 test versions of ObjectSizeCalculator that is using caffeine and guava, the code is at:
Caffeine:
https://gist.github.com/vanhoale/02cf479ae26c04946251cae645bcc776
Guava:
https://gist.github.com/vanhoale/b12324695b85a4d4731618749330fbb3

The tests are to calculate the average processing time in nanoseconds of the method: ObjectSizeCalculator.getObjectSize. The results are below:
ObjectSizeCalculator with Caffeine:
image

ObjectSizeCalculator with Guava:
image

Thoughts/advice?

**Comment by ben-manes** [comment]
Thanks @vanhoale. This makes sense because it is not one-to-one, as you added a workaround to a limitation of `ConcurrentHashMap`. The result being a higher penalty by delegating the compute to a thread pool since CHM does not allow recursive computations. None of the data structures support it, but Guava does work better (it used to deadlock in that case, which I fixed).

```java
// Have to use buildAsync with a synchronous view due to a CHM bug in Java 8.
// See https://github.com/ben-manes/caffeine/issues/89
private final LoadingCache<Class<?>, ClassSizeInfo> classSizeInfos = Caffeine.newBuilder()
        .buildAsync(ClassSizeInfo::new)
        .synchronous();
```

The alternative would be to either use a racy get-load-put call or to implement locking yourself. The racy approach is probably okay since this has no I/O and should be fast. The manual locking isn't too hard, but adds code complexity. Either way, continuing to use Guava is probably the simplest and best choice. I agree with your decision and thanks for clarifying what the slowdown was due to.

**Comment by vanhoale** [comment]
> ```java
> CHM bug
> ```

yeah, it seems to be [fixed](https://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8028564) at jdk8_120. Thanks @ben-manes for the explanation

**Comment by jpechane** [comment]
@vanhoale Please check now the content of installation artifacts. It contains guava (which is expected) and then couple of more probably pulled in by Guava. Are they necessary for runtime? If not could you please exclude them in `debezium-assembly-descriptors/src/main/resources/assemblies/*`?

**Comment by jpechane** [comment]
@vanhoale Squashed and applied, thanks. Could you please also take a look at https://issues.redhat.com/browse/DBZ-2766

I wonder if we could reduce the performance penalty with it or not.

**Comment by vanhoale** [comment]
> @vanhoale Squashed and applied, thanks. Could you please also take a look at https://issues.redhat.com/browse/DBZ-2766
> 
> I wonder if we could reduce the performance penalty with it or not.

thanks @jpechane ! Yes, I can take a look at it!

**Comment by jpechane** [comment]
@vanhoale Hi, I've just noticed that metrics are documented but ot the config option itself. Could you pleas eprepare a new PR with docs update? Thanks

**Pull request merged**